doc/sphinx/proof-engine/ltac.rst

```diff
@@ -1083,45 +1567,49 @@ Testing boolean expressions
    .. exn:: Condition not satisfied.
       :undocumented:
 
-Proving a subgoal as a separate lemma
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Proving a subgoal as a separate lemma: abstract
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-.. tacn:: abstract @ltac_expr
-   :name: abstract
+.. tacn:: abstract @ltac_expr2 {? using @ident }
 
-   From the outside, :n:`abstract @ltac_expr` is the same as :n:`solve @ltac_expr`.
-   Internally it saves an auxiliary lemma called ``ident_subproofn`` where
-   ``ident`` is the name of the current goal and ``n`` is chosen so that this is
-   a fresh name. Such an auxiliary lemma is inlined in the final proof term.
+   Does a :tacn:`solve` :n:`[ @ltac_expr2 ]` and saves the subproof as an auxiliary lemma.
+   if :token:`ident` is specified, the lemma is saved with that name; otherwise
+   the lemma is saved with the name `NAME_subproofN` where
+   `NAME` is the name of the current goal and `N` is
+   a fresh name. The auxiliary lemma is inlined in the final proof term.
```

Comment:
```suggestion
   the lemma is saved with the name :n:`@ident_subproof{? @num }` where
   :n`@ident` is the name of the current goal and :n:`{? @num }` is chosen
   so that the resulting name is fresh. The auxiliary lemma is inlined in the final proof term.
```

Reply:
Details:
- Needed to used ```:n:`@ident`\ `_subproof`\ :n:`{? @num }` ``` here
- I didn't like "name of the current goal" at all--I've not seen that phrase before.  The best I could think of was to add "(e.g. the theorem name)".
- Not sure what "The auxiliary lemma is inlined in the final proof term." means.  If I print `Exists_dec` in `Lists.v`, I see the lemma names but I don't see anything inlined.  `Search "Exists_dec"` doesn't show any lemma--where is it?  Seems like you should be able to see it.

```
Exists_dec = 
fun (l : list A) (Pdec : forall x : A, {P x} + {~ P x}) =>
list_rec (fun l0 : list A => {Exists l0} + {~ Exists l0}) (right Exists_dec_subproof)
  (fun (a : A) (l' : list A) (Hrec : {Exists l'} + {~ Exists l'}) =>
   match Hrec with
   | left Hl' => left (Exists_cons_tl a Hl')
   | right Hl' =>
       let s := Pdec a in
       match s with
       | left Ha => left (Exists_cons_hd l' Ha)
       | right Ha => right (Exists_dec_subproof0 Hl' Ha)
       end
   end) l
     : forall l : list A, (forall x : A, {P x} + {~ P x}) -> {Exists l} + {~ Exists l}

Arguments Exists_dec _%list_scope _%function_scope
```

Reply:
The auxiliary lemma is inlined in the final proof term *only* when the theorem is closed with `Qed`.  Consider:
```coq
Lemma foo : True. Proof. abstract exact I. Qed.
Lemma fooT : True. Proof. abstract exact I. Defined.
Print foo.
(* foo = (fun foo_subproof : True => foo_subproof) I
     : True
 *)
Print fooT.
(* fooT = fooT_subproof
     : True
*)
```

> Search "Exists_dec" doesn't show any lemma--where is it? 

This is because of https://github.com/coq/coq/blob/0ae20176df146b90f11e131aff3673584e6ffb3a/theories/Init/Prelude.v#L59 If you do
```coq
Require Import Coq.Lists.List.
Remove Search Blacklist "_subproof".
Search "Exists_dec".
(* Exists_dec_subproof: forall (A : Type) (P : A -> Prop), ~ Exists P nil
Exists_dec_subproof0:
  forall (A : Type) (P : A -> Prop) (a : A) (l' : list A),
  ~ Exists P l' -> ~ P a -> ~ Exists P (a :: l')
Exists_dec:
  forall (A : Type) (P : A -> Prop) (l : list A),
  (forall x : A, {P x} + {~ P x}) -> {Exists P l} + {~ Exists P l}
Forall_Exists_dec:
  forall (A : Type) (P : A -> Prop),
  (forall x : A, {P x} + {~ P x}) ->
  forall l : list A, {Forall P l} + {Exists (fun x : A => ~ P x) l}
*)
```